### PR TITLE
framework/arastorage: enhance bptree indexing feature

### PR DESCRIPTION
**Pull request by SRCN-LiQin**
add delete function for bplustree algorithm
apply it to tuple remove processing
@Taejun-Kwon @jeongarmy

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [4aa110eea32afe4a873cf3eb3290a4afeb3bbcd4] - Code Rule Check Result:
apps/examples/testcase/ta_tc/arastorage/utc/utc_arastorage_main.c:1487: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/ta_tc/arastorage/utc/utc_arastorage_main.c:1488: ERROR: [IDT_M_TAB] code indent should use tabs where possible
framework/src/arastorage/index_bplustree.c:781: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:782: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:783: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:784: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:784: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)
framework/src/arastorage/index_bplustree.c:2038: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:2069: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
framework/src/arastorage/index_bplustree.c:2082: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:2106: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2304: ERROR: [SPC_M_OPR] spaces required around that '&&' (ctx:WxV)
framework/src/arastorage/index_bplustree.c:2307: ERROR: [SPC_M_OPR] spaces required around that '>=' (ctx:WxV)
framework/src/arastorage/index_bplustree.c:2367: ERROR: [SPC_M_OPR] spaces required around that '?' (ctx:VxW)
framework/src/arastorage/index_bplustree.c:2370: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
framework/src/arastorage/index_bplustree.c:2378: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2398: ERROR: [SPC_M_OPR] spaces required around that '?' (ctx:VxW)
framework/src/arastorage/index_bplustree.c:2399: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)
framework/src/arastorage/index_bplustree.c:2404: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2499: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)
framework/src/arastorage/index_bplustree.c:2519: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2523: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_bplustree.c:2553: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxO)
framework/src/arastorage/index_bplustree.c:2553: ERROR: [SPC_M_OPR] space required before that '&' (ctx:OxV)
framework/src/arastorage/index_bplustree.c:2561: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2566: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_bplustree.c:2580: ERROR: [BRC_M_SMT] else should follow close brace '}'
framework/src/arastorage/index_manager.c:297: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_manager.c:298: ERROR: [IDT_M_TAB] code indent should use tabs where possible
framework/src/arastorage/index_manager.c:298: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/index_manager.c:299: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/arastorage/relation.c:969: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [4aa110eea32afe4a873cf3eb3290a4afeb3bbcd4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [67b9f6855011822a86303a65c7d2536688d51daf] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [67b9f6855011822a86303a65c7d2536688d51daf] - Code Rule Check OK.

File: framework/src/arastorage/index_bplustree.c [hunk 1]
```diff
@@ -2002,3 +2023,565 @@ static int db_flush(tree_t *tree, relation_t *rel)
...
+	for (i = 0; i < bucket->next_free_slot; i++) {
+		if (bucket->pairs[i].key == value) {
+			removed[rm_cnt++] = bucket->pairs[i].value;
+			bucket->next_free_slot--;
+			for (j = i; j < bucket->next_free_slot; j++) {
+				bucket->pairs[j].key = bucket->pairs[j + 1].key;
+				bucket->pairs[j].value = bucket->pairs[j + 1].value;
+			}
+
+			if (remove_all == 0) break; //delete only one pair, make it possible to update
```
[the remark this answers is not in the document]
Comment (SRCN-LiQin) [reply framework/src/arastorage/index_bplustree.c]: All Fixed, thanks~

**Comment by seinfra** [comment]
Target : [43e36c134617ce760538b70206e8623a4c4d8585] -

**Comment by seinfra** [comment]
Target : [0294bf3381e09c1e2f3a715d8df1e8e95c2da077] - Code Rule Check Result:
framework/src/arastorage/index_bplustree.c:2483: ERROR: [IDT_M_TAB] please, no space before tabs

**Comment by seinfra** [comment]
Target : [0294bf3381e09c1e2f3a715d8df1e8e95c2da077] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/arastorage/utc/utc_arastorage_main.c [hunk 2]
```diff
@@ -1435,6 +1435,84 @@ static void utc_arastorage_cursor_get_string_value_n(void)
...
+	snprintf(query, QUERY_LENGTH, "CREATE ATTRIBUTE %s DOMAIN int IN %s;", g_attribute_set[3], RELATION_NAME1);
+	res = db_exec(query);
+	TC_ASSERT_EQ("db_exec", DB_SUCCESS(res), true);
+
+	/*=========================INSERT TUPLE==========================*/
+	for (i = 0; i < 240; i++) {
+		memset(query, 0, QUERY_LENGTH);
+		snprintf(query, QUERY_LENGTH, "INSERT (%d, %ld, \'%s\', %d) INTO %s;",
+			i, g_arastorage_data_set[i%DATA_SET_NUM].long_value,
+			g_arastorage_data_set[i%DATA_SET_NUM].string_value, /*1000 - i*/rand()%500, RELATION_NAME1);
```
Comment (xiarihf) [inline apps/examples/testcase/ta_tc/arastorage/utc/utc_arastorage_main.c]: space needs around %

File: framework/src/arastorage/index_bplustree.c [hunk 3]
```diff
@@ -318,6 +319,13 @@ int compare(const void *p1, const void *p2)
 	return 1;
 }
 
+void* bptree_malloc(size_t size)
```
Comment (xiarihf) [inline framework/src/arastorage/index_bplustree.c]: why not using zalloc？and the function should be named bptree_zalloc?

File: framework/src/arastorage/index_bplustree.c [hunk 4]
```diff
@@ -2002,3 +2023,560 @@ static int db_flush(tree_t *tree, relation_t *rel)
...
+	}
+
+	modify_cache(tree, node_id, NODE, UNLOCK);
+	modify_cache(tree, pnode_id, NODE, UNLOCK);
+
+	return 0;
+}
+
+static int bucket_merge_sibling(tree_t *tree, pair_t *path, bucket_t *bucket, int rm_val)
+{
```
Comment (xiarihf) [inline framework/src/arastorage/index_bplustree.c]: No input pointer check. if "tree" is null, it will crash.

Comment (SRCN-LiQin) [reply framework/src/arastorage/index_bplustree.c]: it's internal function which could guarantee this

File: framework/src/arastorage/relation.c [hunk 5]
```diff
@@ -991,8 +1015,24 @@ db_result_t relation_process_remove(db_handle_t **handle, db_cursor_t *cursor)
...
+			}
+			from_attr = from_attr->next;
+		}
+	}
+
 	/* Check whether the given predicate is true for this tuple. */
-	if ((*handle)->lvm_instance == NULL || lvm_execute((*handle)->lvm_instance) == FALSE) {
+	if (not_remove) {
 		result = storage_put_row((*handle)->result_rel, result_row, TRUE);
 		if (DB_ERROR(result)) {
```
Comment (jeongarmy) [inline framework/src/arastorage/relation.c]: If a index is added in line 1026 and then this error happens, index information is wrong,
I think you need to handle this case.

File: framework/src/arastorage/relation.c [hunk 6]
```diff
@@ -315,11 +315,32 @@ relation_t *relation_create(char *name, db_direction_t dir)
 	return NULL;
 }
 
-db_result_t relation_rename(char *old_name, char *new_name)
+
+//remove relation file and tuple file, but not index file. for tuple remove operation
+db_result_t relation_tuple_remove(relation_t *rel, int remove_tuples)
```
Comment (jeongarmy) [inline framework/src/arastorage/relation.c]: I think a function name looks ambiguous.

File: framework/src/arastorage/relation.c [hunk 7]
```diff
@@ -1023,6 +1063,9 @@ db_result_t relation_process_remove(db_handle_t **handle, db_cursor_t *cursor)
 		goto errout;
 	}
 	memcpy((*handle)->result_rel->name, (*handle)->rel->name, sizeof((*handle)->result_rel->name));
+	if (storage_get_row_amount((*handle)->result_rel, &nrows) == DB_OK) {
```
Comment (jeongarmy) [inline framework/src/arastorage/relation.c]: Do you use storage_get_row_amount to verify to write well? It not, result_rel->cardinality is enough.

File: framework/src/arastorage/index_bplustree.c [hunk 8]
```diff
@@ -400,13 +412,18 @@ static db_result_t create(index_t *index)
...
-	buck.info[0] = CONFIG_BUCKETS_LIMIT - 1;
-	buck.info[1] = KEY_MAX;
-	buck.info[2] = 0;
-	storage_write_to(tree->bucket_storage, &buck, 0, sizeof(bucket_t));
-
+	tree_node = bptree_malloc(tree_node_size);
+	storage_write_to(tree->tree_storage, tree_node, offset, tree_node_size);
+	free(tree_node);
+
+	buck = bptree_malloc(buck_size);
```
Comment (jeongarmy) [inline framework/src/arastorage/index_bplustree.c]: need checking NULL

File: framework/src/arastorage/index_bplustree.c [hunk 9]
```diff
@@ -2002,3 +2023,560 @@ static int db_flush(tree_t *tree, relation_t *rel)
...
+	if (tree->lock_buckets[bucket_id] == 0) {
+		DB_LOG_E("PANIC EDITED BUCKET WITHOUT LOCK\n");
+		pthread_mutex_unlock(&(tree->bucket_lock));
+		rw_unlock_write(&(tree->tree_lock));
+		ret = DB_INDEX_ERROR;
+	} else {
+		tree->lock_buckets[bucket_id] = 0;
+		pthread_mutex_unlock(&(tree->bucket_lock));
+		rw_unlock_write(&(tree->tree_lock));
+	}
```
Comment (jeongarmy) [inline framework/src/arastorage/index_bplustree.c]: free(path) is missed

File: framework/src/arastorage/index_bplustree.c (hunk 8 above)
Comment (SRCN-LiQin) [reply framework/src/arastorage/index_bplustree.c]: fixed, add NULL checking

File: framework/src/arastorage/index_bplustree.c [hunk 10]
```diff
@@ -318,6 +319,13 @@ int compare(const void *p1, const void *p2)
 	return 1;
 }
 
+void* bptree_malloc(size_t size)
+{
+	void *p = malloc(size);
```
[the remark this answers is not in the document]
Comment (SRCN-LiQin) [reply framework/src/arastorage/index_bplustree.c]: fixed, add checking before memset.

File: framework/src/arastorage/relation.c (hunk 7 above)
Comment (SRCN-LiQin) [reply framework/src/arastorage/relation.c]: use to do lseek, avoid occasional row amout reading issue after save file.

File: framework/src/arastorage/relation.c (hunk 6 above)
Comment (SRCN-LiQin) [reply framework/src/arastorage/relation.c]: refined new function name

File: framework/src/arastorage/relation.c (hunk 5 above)
Comment (SRCN-LiQin) [reply framework/src/arastorage/relation.c]: refined the code logic

**Comment by seinfra** [comment]
Target : [a8ffa4b6177935d1c42fde53bd8d595565e7b9b7] -

**Comment by seinfra** [comment]
Target : [a8ffa4b6177935d1c42fde53bd8d595565e7b9b7] -

**Comment by seinfra** [comment]
Target : [ecf530d9213a08cf06c93244534c06e7a92d3148] - Code Rule Check Result:
framework/src/arastorage/index_bplustree.c:2489: ERROR: [IDT_M_TAB] please, no space before tabs

**Comment by seinfra** [comment]
Target : [ecf530d9213a08cf06c93244534c06e7a92d3148] - Code Rule Check (C++) OK.

File: framework/src/arastorage/relation.c [hunk 11]
```diff
@@ -317,9 +317,8 @@ relation_t *relation_create(char *name, db_direction_t dir)
 
 db_result_t relation_rename(char *old_name, char *new_name)
 {
-	relation_t *rel = relation_load(new_name);
-
-	if (DB_ERROR(relation_remove(rel, 1)) || DB_ERROR(storage_rename_relation(old_name, new_name))) {
+	DB_LOG_D("old_name is %s, new_name is %s\n", old_name, new_name);
+	if (DB_ERROR(storage_rename_relation(old_name, new_name))) {
```
Comment (Taejun-Kwon) [inline framework/src/arastorage/relation.c]: good.

File: framework/src/arastorage/relation.c [hunk 12]
```diff
@@ -705,6 +704,48 @@ static void relation_index_clear(relation_t *rel)
 	free(filename);
 }
 
+static void relation_tuple_index_delete(db_handle_t **handle, unsigned char *row_ptr, int update_index)
```
Comment (Taejun-Kwon) [inline framework/src/arastorage/relation.c]: How about relation_delete_index_item?

**Comment by seinfra** [comment]
Target : [89cb10ca4d71978ce874f7eb66dc511e997b5454] -

**Comment by seinfra** [comment]
Target : [89cb10ca4d71978ce874f7eb66dc511e997b5454] -

**Comment by seinfra** [comment]
Target : [8d800af7ab1908d5e46cc07be71ad1f94f17e97c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8d800af7ab1908d5e46cc07be71ad1f94f17e97c] - Code Rule Check OK.

**Pull request merged**